File: Binary Trees/tree-includes.js
class Node {
  constructor(val) {
    this.val = val;
    this.left = null;
    this.right = null;
  }
}

// iterative breadth first
// const treeIncludes = (root, target) => {
//   if (root === null) return false;

//   const queue = [root];
//   while (queue.length > 0) {
//     const current = queue.shift();
//     if (current.val === target) return true;

//     if (current.left) queue.push(current.left);
//     if (current.right) queue.push(current.right);
//   }
//   return false;
// };

// recursive depth first
const treeIncludes = (root, target) => {
  if (root === null) return false;
  if (root.val === target) return true;
  return treeIncludes(root.left, target) || treeIncludes(root.right, target);
};

module.exports = {
  Node,
  treeIncludes,
};

File: Binary Trees/tree_includes.py
class Node: 
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


# def tree_includes(root, target): 
#     if root is None:
#         return False

#     queue = [root] 
    
#     while queue:
#         current = queue.pop()
#         if current.val is target:
#             return True
        
#         if current.left is not None: 
#             queue.append(current.left)
#         if current.right is not None: 
#             queue.append(current.right)
    
#     return False

def tree_includes(root, target): 
    if root is None:
        return False
    
    if root.val is target:
        return True 

    return tree_includes(root.left, target) or tree_includes(root.right, target)
    